### ISSUE TITLE
 No Activity found to handle Intent { act=android.intent.action.VIEW cat=[android.intent.category.BROWSABLE] 

### ISSUE THREAD

**Issue by kensykora** [AzureAD/microsoft-authentication-library-for-dotnet#547]
We picked up this crash report form HockeyApp in our production builds from our users. It seems to be originating from the package. Login and Registration works normally on the Android platform for us so this seems to be some special circumstance that I'm unclear of why it might occur (Activity was recycled somehow during login/registration?)

Android Version: 7.0
Microsoft.Identity.Client: 1.1.2-preview0008
Xamarin.Forms: 2.5.0.280555
Xamarin.android.Support.v4: 26.1.0.1

```
android.content.ActivityNotFoundException: No Activity found to handle Intent { act=android.intent.action.VIEW cat=[android.intent.category.BROWSABLE] dat=https://login.microsoftonline.com/... pkg=com.android.chrome launchParam=MultiScreenLaunchParams { mDisplayId=0 mFlags=0 } }
android.app.Instrumentation.checkStartActivityResult()Instrumentation.java:1839
android.app.Instrumentation.execStartActivity()Instrumentation.java:1531
android.app.Activity.startActivityForResult()Activity.java:4399
android.app.Activity.startActivityForResult()Activity.java:4358
android.app.Activity.startActivity()Activity.java:4682
android.app.Activity.startActivity()Activity.java:4650
microsoft.identity.client.AuthenticationActivity.n_onResume(Native Method)
microsoft.identity.client.AuthenticationActivity.onResume()AuthenticationActivity.java:48
android.app.Instrumentation.callActivityOnResume()Instrumentation.java:1277
android.app.Activity.performResume()Activity.java:7088
android.app.ActivityThread.performResumeActivity()ActivityThread.java:3768
android.app.ActivityThread.handleResumeActivity()ActivityThread.java:3832
android.app.ActivityThread.handleLaunchActivity()ActivityThread.java:2994
at android.app.ActivityThread.-wrap14(ActivityThread.java)
android.app.ActivityThread$H.handleMessage()ActivityThread.java:1631
android.os.Handler.dispatchMessage()Handler.java:102
android.os.Looper.loop()Looper.java:154
android.app.ActivityThread.main()ActivityThread.java:6682
java.lang.reflect.Method.invoke(Native Method)
com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run()ZygoteInit.java:1534
com.android.internal.os.ZygoteInit.main()ZygoteInit.java:1424
Xamarin caused by: Android.Content.ActivityNotFoundException: No Activity found to handle Intent { act=android.intent.action.VIEW cat=[android.intent.category.BROWSABLE] dat=https://login.microsoftonline.com/... pkg=com.android.chrome launchParam=MultiScreenLaunchParams { mDisplayId=0 mFlags=0 } }
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()<0fd49e8d72c3425a9142b62bf7bb5408>:0
Java.Interop.JniEnvironment.InstanceMethods.CallNonvirtualVoidMethod(JniObjectReference instance, JniObjectReference type, JniMethodInfo method, JniArgumentValue* args)<7bdf46b1ae9e4b5dbe64b85267ccc954>:0
Java.Interop.JniPeerMembers.JniInstanceMethods.InvokeVirtualVoidMethod(string encodedMember, IJavaPeerable self, JniArgumentValue* parameters)<7bdf46b1ae9e4b5dbe64b85267ccc954>:0
Android.Content.ContextWrapper.StartActivity(Intent intent)<7eefb6695f9c4eb7ae386b276deee35f>:0
Microsoft.Identity.Client.AuthenticationActivity.OnResume()<1a1ee4c087634a4da11ac7e4aec10a38>:0
Android.App.Activity.n_OnResume(IntPtr jnienv, IntPtr native__this)<7eefb6695f9c4eb7ae386b276deee35f>:0
at (wrapper dynamic-method) System.Object.dc10c0e9-8c32-4e82-8a0d-868e3b0d0233(intptr,intptr)
--- End of managed Android.Content.ActivityNotFoundException stack trace ---
android.content.ActivityNotFoundException: No Activity found to handle Intent { act=android.intent.action.VIEW cat=[android.intent.category.BROWSABLE] dat=https://login.microsoftonline.com/... pkg=com.android.chrome launchParam=MultiScreenLaunchParams { mDisplayId=0 mFlags=0 } }
android.app.Instrumentation.checkStartActivityResult()Instrumentation.java:1839
android.app.Instrumentation.execStartActivity()Instrumentation.java:1531
android.app.Activity.startActivityForResult()Activity.java:4399
android.app.Activity.startActivityForResult()Activity.java:4358
android.app.Activity.startActivity()Activity.java:4682
android.app.Activity.startActivity()Activity.java:4650
microsoft.identity.client.AuthenticationActivity.n_onResume(Native Method)
microsoft.identity.client.AuthenticationActivity.onResume()AuthenticationActivity.java:48
android.app.Instrumentation.callActivityOnResume()Instrumentation.java:1277
android.app.Activity.performResume()Activity.java:7088
android.app.ActivityThread.performResumeActivity()ActivityThread.java:3768
android.app.ActivityThread.handleResumeActivity()ActivityThread.java:3832
android.app.ActivityThread.handleLaunchActivity()ActivityThread.java:2994
at android.app.ActivityThread.-wrap14(ActivityThread.java)
android.app.ActivityThread$H.handleMessage()ActivityThread.java:1631
android.os.Handler.dispatchMessage()Handler.java:102
android.os.Looper.loop()Looper.java:154
android.app.ActivityThread.main()ActivityThread.java:6682
java.lang.reflect.Method.invoke(Native Method)
com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run()ZygoteInit.java:1534
com.android.internal.os.ZygoteInit.main()ZygoteInit.java:1424
```

The interesting lines in this stack trace seem to be:

```
android.app.Activity.startActivity()Activity.java:4650
microsoft.identity.client.AuthenticationActivity.n_onResume(Native Method)
microsoft.identity.client.AuthenticationActivity.onResume()AuthenticationActivity.java:48
android.app.Instrumentation.callActivityOnResume()Instrumentation.java:1277
```

We have a dev and a prod instance of Azure B2C and are registering the Activity Intent filters in our `AndroidManifest.xml` as such (guids replaced with fake IDs, but they are different):

```xml
<manifest xmlns:android="http://schemas.android.com/apk/res/android" package="com.myapp" android:installLocation="auto" android:icon="@drawable/icon" android:versionCode="99999" android:versionName="0.0.0">
	<uses-sdk android:minSdkVersion="21" android:targetSdkVersion="26" />
	<application android:label="MyApp" android:icon="@drawable/icon">
		<activity android:name="microsoft.identity.client.BrowserTabActivity">
                         <!-- Dev -->
			<intent-filter>
				<action android:name="android.intent.action.VIEW" />
				<category android:name="android.intent.category.DEFAULT" />
				<category android:name="android.intent.category.BROWSABLE" />
				<data android:scheme="msal00000000-0000-0000-0000-000000000000" android:host="auth" />
			</intent-filter>
                         <!-- Prod -->
			<intent-filter>
				<action android:name="android.intent.action.VIEW" />
				<category android:name="android.intent.category.DEFAULT" />
				<category android:name="android.intent.category.BROWSABLE" />
				<data android:scheme="msal00000000-0000-0000-0000-000000000001" android:host="auth" />
			</intent-filter>
		</activity>
	</application>
</manifest>
```

**Comment by jennyf19** [comment]
Thanks @kensykora 
Can you try adding this line to your `AndroidManifest.xml`?
```	<uses-permission android:name="android.permission.INTERNET" />```

**Comment by kensykora** [comment]
Hey @jennyf19 we have both of these permissions in our manifest that is generated through assembly attributes with Xamarins build process, it’s not specified in the xml included here. I can post the full generated manifest file if it would be helpful

**Comment by jennyf19** [comment]
@kensykora - Did the activity reject just start happening?

**Comment by kensykora** [comment]
That’s interesting, thanks for the heads up. Any idea on when you might expect a fix would be appreciated.

**Comment by jennyf19** [comment]
@kensykora We're working on, and implementing, the design right now. We'll let you know when we have an update that you could test out. Not giving a specific date at this point though :)

**Comment by kensykora** [comment]
Just FYI, we figured out reliable steps to reproduce. This seems to occur when users disable the Chrome browser app in their settings.

Steps to reproduce:

1. Navigate to settings > Apps 
2. Select Chrome
3. Tap "Disable" button, confirm warnings
4. Launch an app with this component integrated on Android.
5. Attempt to start a login workflow

Expected: ?? Some other default browser is used instead ? Fallback to web view?
Actual: Application crashes

**Comment by jmprieur** [comment]
+ @jennyf19 who is working on fixing this experience;
Indeed, for the moment, on Xamarin.Android, MSAL.NET requires Chrome tabs to be present. We are working on fixing this issue.

**Comment by AnandPhadke** [comment]
I am facing same issue, how to use  msal library without chrome ?

**Comment by jmprieur** [comment]
you cannot today, but @jennyf19 has the fix.

**Comment by jennyf19** [comment]
@AnandPhadke 
The Webview changes have been implemented for MSAL in the combined library repo, which is inside the [ADAL repo](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet). You are welcome to try out MSAL on the dev branch and get back to us with feedback and suggestions. 

We are still working out the implementation details, but for the time being, you can use this as a [guide](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/wiki/Webviews) to try out the [webview options](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/wiki/Webviews#msalnet) for MSAL.

You can also try out the [Xamarin Android](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/dev/msal/tests/dev%20apps/XForms/XForms.Android/MainActivity.cs#L50) app or the [Xamarin iOS](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/dev/msal/tests/dev%20apps/XForms/XForms.iOS/AppDelegate.cs#L55) app in MSAL.

When we release the updated webview option for MSAL, it will be from the ADAL repo, more details to follow. We don't have a date for the release, as we are waiting on a few more features for MSAL. You are welcome to give us feedback, as we have not closed on how it will be surfaced to the developer. 

Email me at jeferrie@microsoft.com if you want.

**Comment by AnandPhadke** [comment]
I am working on the native android app, I don't have chrome install in my device(customise OS ). Can you please help me to use the library without chrome either with default browser or webview. 
Thanks in advance!!

**Comment by jmprieur** [comment]
@AnandPhadke: if you are working on a native android app, then this library (ADAL.Net) is probably not what you are looking for. You should ask your question in https://github.com/AzureAD/microsoft-authentication-library-for-android/issues

**Comment by jmprieur** [comment]
@AnandPhadke : closing it. Feel free to reopen if this is still an issue for you

**Comment by hearbeathorror** [comment]
@jennyf19, @jmprieur .. not sure if I am posting in the right issue or branch but I wanted to know if it possible with the medal android SDK to open a custom webview or embedded webview instead of the Chrome browser for authentication. ? Or is that still under development? Kindly let me know please. Thanks

**Comment by jennyf19** [comment]
@hearbeathorror You can use embedded webview in the current version of MSAL. Here is information on the [embedded webview](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/MSAL.NET-uses-web-browser#you-can-also-enable-embedded-webviews-in-xamarinios-and-xamarinandroid-apps) and [system browser on androd](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/Android-system-browser) and how to set up the [UIParent](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/Acquiring-tokens-interactively#controlling-the-location-of-the-dialog-with-the-parent-parameters-uiparent). Please note that for b2c, when using the new b2clogin.com authority, embedded webview will not work with [Google Auth](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/AAD-B2C-specificities#google-auth-and-embedded-webview).

**Comment by hearbeathorror** [comment]
Thank you for these.. I did go through them, but then i got confused. So
will this work on native android? Or MSAL-.net is a different app
altogether? I mean can i do this from native android, or set it in native
android to use the embedded webview? Or do i need a backend team to do it
for me? Or the portal or something of the sort?

**Comment by jennyf19** [comment]
@hearbeathorror MSAL .NET uses xamarin to work cross-platform on android and ios, but you can also use our native [android library](https://github.com/AzureAD/microsoft-authentication-library-for-android) as well.

**Comment by hearbeathorror** [comment]
@jennyf19 , i did get a way to let the android sdk open a webview instead of using the browser. However, i face another issue: https://github.com/AzureAD/microsoft-authentication-library-for-android/issues/305, is there some one who could assist? I do not know if the android team is active on github. 

Thanks

**Comment by jennyf19** [comment]
@hearbeathorror I'll ping the android team.
